# Creating TCGA cohorts  (part 1)

This notebook will show you how to create a TCGA cohort using the publicly available TCGA BigQuery tables that the [ISB-CGC](http://isb-cgc.org) project has produced based on the open-access [TCGA](http://cancergenome.nih.gov/) data available at the [Data Portal](https://tcga-data.nci.nih.gov/tcga/).  You will need to have access to a Google Cloud Platform (GCP) project in order to use BigQuery.  If you don't already have one, you can sign up for a [free-trial](https://cloud.google.com/free-trial/) or contact [us](mailto://info@isb-cgc.org) and become part of the community evaluation phase of our [Cancer Genomics Cloud pilot](https://cbiit.nci.nih.gov/ncip/nci-cancer-genomics-cloud-pilots).

We are not attempting to provide a thorough BigQuery or IPython tutorial here, as a wealth of such information already exists.  Here are some links to some resources that you might find useful: 
* [BigQuery](https://cloud.google.com/bigquery/what-is-bigquery), 
* the BigQuery [web UI](https://bigquery.cloud.google.com/) where you can run queries interactively, 
* [IPython](http://ipython.org/) (now known as [Jupyter](http://jupyter.org/)), and 
* [Cloud Datalab](https://cloud.google.com/datalab/) the recently announced interactive cloud-based platform that this notebook is being developed on. 

There are also many tutorials and samples available on github (see, in particular, the [datalab](https://github.com/GoogleCloudPlatform/datalab) repo and the [Google Genomics](  https://github.com/googlegenomics) project).

OK then, let's get started!  In order to work with BigQuery, the first thing you need to do is import the bigquery module:

In [1]:
import gcp.bigquery as bq

The next thing you need to know is how to access the specific tables you are interested in.  BigQuery tables are organized into datasets, and datasets are owned by a specific GCP project.  The tables we will be working with in this notebook are in a dataset called **`tcga_201510_alpha`**, owned by the **`isb-cgc`** project.  A full table identifier is of the form `<project_id>:<dataset_id>.<table_id>`.  Let's start by getting some basic information about the tables in this dataset:

In [2]:
d = bq.DataSet('isb-cgc:tcga_201510_alpha')
for t in d.tables():
  print '%10d rows  %12d bytes   %s' \
      % (t.metadata.rows, t.metadata.size, t.name.table_id)

     17519 rows       6072735 bytes   Annotations
     23688 rows       6361090 bytes   Biospecimen_data
     11152 rows       4215264 bytes   Clinical_data
   2586033 rows     326181970 bytes   Copy_Number_segments
3874187076 rows  437382491205 bytes   DNA_Methylation_betas
   1618973 rows     180121639 bytes   Protein_RPPA_data
   5813281 rows    6392929246 bytes   Somatic_Mutation_calls
  38299138 rows    4459086535 bytes   mRNA_BCGSC_HiSeq_RPKM
 212475319 rows   25083953060 bytes   mRNA_UNC_HiSeq_RSEM
  13483548 rows    1644762154 bytes   miRNA_expression


In this tutorial, we are going to look at a few different ways that we can use the information in these tables to create cohorts.  Now, you maybe asking what we mean by "cohort" and why you might be interested in *creating* one, or maybe what it even means to "create" a cohort.  The TCGA dataset includes clinical, biospecimen, and molecular data from over 10,000 cancer patients who agreed to be a part of this landmark research project to build [The Cancer Genome Atlas](http://cancergenome.nih.gov/).  This large dataset was originally organized and studied according to [cancer type](http://cancergenome.nih.gov/cancersselected) but now that this multi-year project is nearing completion, with over 30 types of cancer and over 10,000 tumors analyzed, **you** have the opportunity to look at this dataset from whichever angle most interests you.  Maybe you are particularly interested in early-onset cancers, or gastro-intestinal cancers, or a specific type of genetic mutation.  This is where the idea of a "cohort" comes in.  The original TCGA "cohorts" were based on cancer type (aka "study"), but now you can define a cohort based on virtually any clinical or molecular feature by querying these BigQuery tables.  A cohort is simply a list of samples, using the [TCGA barcode](https://wiki.nci.nih.gov/display/TCGA/TCGA+barcode) system.  Once you have created a cohort you can use it in any number of ways: you could further explore the data available for one cohort, or compare one cohort to another, for example.

In the rest of this tutorial, we will create several different cohorts based on different motivating research questions.  We hope that these examples will provide you with a starting point from which you can build, to answer your own research questions.

### Exploring the Clinical data table
Let's start by looking at the clinical data table.  The TCGA dataset contains a few very basic clinical data elements for almost all patients, and contains additional information for some tumor types only.  For example smoking history information is generally available only for lung cancer patients, and BMI (body mass index) is only available for tumor types where that is a known significant risk factor.  Let's take a look at the clinical data table and see how many different pieces of information are available to us:

In [3]:
%bigquery schema --table isb-cgc:tcga_201510_alpha.Clinical_data

That's a lot of fields!  We can also get at the schema programmatically:

In [4]:
table = bq.Table('isb-cgc:tcga_201510_alpha.Clinical_data')
if ( table.exists() ):
    fieldNames = map(lambda tsf: tsf.name, table.schema)
    fieldTypes = map(lambda tsf: tsf.data_type, table.schema)
    print " This table has %d fields. " % ( len(fieldNames) )
    print " The first few field names and types are: " 
    print "     ", fieldNames[:5]
    print "     ", fieldTypes[:5]
else: 
    print " There is no existing table called %s:%s.%s" % ( table.name.project_id, table.name.dataset_id, table.name.table_id )

 This table has 65 fields. 
 The first few field names and types are: 
      [u'ParticipantBarcode', u'Study', u'Project', u'ParticipantUUID', u'TSSCode']
      [u'STRING', u'STRING', u'STRING', u'STRING', u'STRING']


Let's look at these fields and see which ones might be the most "interesting", by looking at how many times they are filled-in (not NULL), or how much variation exists in the values.  If we wanted to look at just a single field, "tobacco_smoking_history" for example, we could use a very simple query to get a basic summary:

In [5]:
%%sql 

SELECT tobacco_smoking_history, COUNT(*) AS n
FROM [isb-cgc:tcga_201510_alpha.Clinical_data]
GROUP BY tobacco_smoking_history
ORDER BY n DESC

QueryResultsTable job_uViwLTT9hxIE_JJR-Vi4gkvqghk

But if we want to loop over all 65 fields and get a sense of which fields might provide us with useful criteria for specifying a cohort, we'll want to automate that.  We'll put a threshold on the minimum number of patients that we expect information for, and the maximum number of unique values (since fields such as the "ParticipantBarcode" will be unique for every patient and, although we will need that field later, it's probably not useful for defining a cohort).

In [6]:
numPatients = table.metadata.rows
print " The %s table describes a total of %d patients. " % ( table.name.table_id, numPatients )

# let's set a threshold for the minimum number of values that a field should have,
# and also the maximum number of unique values
minNumPatients = int(numPatients*0.80)
maxNumValues = 50

numInteresting = 0
iList = []
for iField in range(len(fieldNames)):
  aField = fieldNames[iField]
  aType = fieldTypes[iField]
  try:
    qString = "SELECT {0} FROM [{1}]".format(aField,table)
    query = bq.Query(qString)
    df = query.to_dataframe()
    summary = df[str(aField)].describe()
    if ( aType == "STRING" ):
      topFrac = float(summary['freq'])/float(summary['count'])
      if ( summary['count'] >= minNumPatients ):
        if ( summary['unique'] <= maxNumValues and summary['unique'] > 1 ):
          if ( topFrac < 0.90 ):
            numInteresting += 1
            iList += [aField]
            print "     > %s has %d values with %d unique (%s occurs %d times) " \
              % (str(aField), summary['count'], summary['unique'], summary['top'], summary['freq'])
    else:
      if ( summary['count'] >= minNumPatients ):
        if ( summary['std'] > 0.1 ):
          numInteresting += 1
          iList += [aField]
          print "     > %s has %d values (mean=%.0f, sigma=%.0f) " \
            % (str(aField), summary['count'], summary['mean'], summary['std'])
  except:
    pass

print " "
print " Found %d potentially interesting features: " % numInteresting
print "   ", iList

 The Clinical_data table describes a total of 11152 patients. 
     > Study has 11152 values with 33 unique (BRCA occurs 1097 times) 
     > age_at_initial_pathologic_diagnosis has 11102 values (mean=59, sigma=14) 
     > batch_number has 11152 values (mean=203, sigma=134) 
     > vital_status has 11148 values with 2 unique (Alive occurs 7548 times) 
     > days_to_birth has 11034 values (mean=-21761, sigma=5268) 
     > days_to_last_known_alive has 11100 values (mean=1033, sigma=1040) 
     > gender has 11152 values with 2 unique (FEMALE occurs 5811 times) 
     > year_of_initial_pathologic_diagnosis has 11023 values (mean=2008, sigma=4) 
     > person_neoplasm_cancer_status has 10254 values with 2 unique (TUMOR FREE occurs 6531 times) 
     > race has 9828 values with 5 unique (WHITE occurs 8179 times) 
 
 Found 10 potentially interesting features: 
    [u'Study', u'age_at_initial_pathologic_diagnosis', u'batch_number', u'vital_status', u'days_to_birth', u'days_to_last_known_alive', 

The above helps us narrow down on which fields are likely to be the most useful, but if you have a specific interest, for example in menopause or HPV status, you can still look at those in more detail very easily: 

In [7]:
%%sql
SELECT menopause_status, COUNT(*) AS n
FROM [isb-cgc:tcga_201510_alpha.Clinical_data]
WHERE menopause_status IS NOT NULL
GROUP BY menopause_status
ORDER BY n DESC

QueryResultsTable job_lC1ltcDJrw5o8GpKxv45Xe3qLDk

We might wonder which specific tumor types have menopause information:

In [8]:
%%sql
SELECT Study, COUNT(*) AS n
FROM [isb-cgc:tcga_201510_alpha.Clinical_data]
WHERE menopause_status IS NOT NULL
GROUP BY Study
ORDER BY n DESC

QueryResultsTable job_dEg13gnxcRymTc88T72R4za2Lqo

In [9]:
%%sql
SELECT hpv_status, hpv_calls, COUNT(*) AS n
FROM [isb-cgc:tcga_201510_alpha.Clinical_data]
WHERE hpv_status IS NOT NULL
GROUP BY hpv_status, hpv_calls
HAVING n > 20
ORDER BY n DESC

QueryResultsTable job_PaFO7w-55hHHDy3CjKTVaPkZxxM

### TCGA Annotations

An additional factor to consider, when creating a cohort is that there may be additional information that might lead one to exclude a particular patient from a cohort.  In certain instances, patients have been redacted or excluded from analyses for reasons such as prior treatment, etc, but since different researchers may have different criteria for using or excluding certain patients or certain samples from their analyses, in many cases the data is still available while at the same time "annotations" may have been entered into a searchable [database](https://tcga-data.nci.nih.gov/annotations/).  These annotations have also been uploaded into a BigQuery table and can be used in conjuction with the other BigQuery tables.

### Early-onset Breast Cancer

Now that we have a better idea of what types of information is available in the Clinical data table, let's create a cohort consisting of female breast-cancer patients, diagnosed at the age of 50 or younger.

In this next code cell, we define several queries within a **`module`** which allows us to use them both individually and by reference in the final, main query.  
+ the first query, called **`select_on_annotations`**, finds all patients in the Annotations table which have either been 'redacted' or had 'unacceptable prior treatment';  
+ the second query, **`select_on_clinical`** selects all female breast-cancer patients who were diagnosed at age 50 or younger, while also pulling out a few additional fields that might be of interest;  and
+ the final query joins these two together and returns just those patients that meet the clinical-criteria and do **not** meet the exclusion-criteria.

In [10]:
%%sql --module createCohort_and_checkAnnotations

DEFINE QUERY select_on_annotations
SELECT
  ParticipantBarcode,
  annotationCategoryName AS categoryName,
  annotationClassification AS classificationName
FROM
  [isb-cgc:tcga_201510_alpha.Annotations]
WHERE
  ( itemTypeName="Patient"
    AND (annotationCategoryName="History of unacceptable prior treatment related to a prior/other malignancy"
      OR annotationClassification="Redaction" ) )
GROUP BY
  ParticipantBarcode,
  categoryName,
  classificationName

DEFINE QUERY select_on_clinical
SELECT
  ParticipantBarcode,
  vital_status,
  days_to_last_known_alive,
  ethnicity,
  histological_type,
  menopause_status,
  race
FROM
  [isb-cgc:tcga_201510_alpha.Clinical_data]
WHERE
  ( Study="BRCA"
    AND age_at_initial_pathologic_diagnosis<=50
    AND gender="FEMALE" )

SELECT
  c.ParticipantBarcode AS ParticipantBarcode
FROM (
  SELECT
    a.categoryName,
    a.classificationName,
    a.ParticipantBarcode,
    c.ParticipantBarcode,
  FROM ( $select_on_annotations ) AS a
  OUTER JOIN EACH 
       ( $select_on_clinical ) AS c
  ON
    a.ParticipantBarcode = c.ParticipantBarcode
  WHERE
    (a.ParticipantBarcode IS NOT NULL
      OR c.ParticipantBarcode IS NOT NULL)
  ORDER BY
    a.classificationName,
    a.categoryName,
    a.ParticipantBarcode,
    c.ParticipantBarcode )
WHERE
  ( a.categoryName IS NULL
    AND a.classificationName IS NULL
    AND c.ParticipantBarcode IS NOT NULL )
ORDER BY
  c.ParticipantBarcode

Here we explicitly call just the first query in the module, and we get a list of 212 patients with one of these disqualifying annotations:

In [11]:
bq.Query(createCohort_and_checkAnnotations.select_on_annotations).results().to_dataframe()

ParticipantBarcode                                       categoryName  \
0         TCGA-BH-A1F5  History of unacceptable prior treatment relate...   
1         TCGA-AR-A2LR  History of unacceptable prior treatment relate...   
2         TCGA-BG-A0MS  History of unacceptable prior treatment relate...   
3         TCGA-BH-A0B6  History of unacceptable prior treatment relate...   
4         TCGA-XK-AAK1  History of unacceptable prior treatment relate...   
5         TCGA-BG-A0M8  History of unacceptable prior treatment relate...   
6         TCGA-T1-A6J8  History of unacceptable prior treatment relate...   
7         TCGA-HT-A619  History of unacceptable prior treatment relate...   
8         TCGA-5L-AAT1  History of unacceptable prior treatment relate...   
9         TCGA-06-6391  History of unacceptable prior treatment relate...   
10        TCGA-BA-4075  History of unacceptable prior treatment relate...   
11        TCGA-IC-A6RF  History of unacceptable prior treatment relate...   
12        TCGA-2G-AAFE  History of unacceptable prior treatment relate...   
13        TCGA-96-A4JK  History of unacceptable prior treatment relate...   
14        TCGA-EJ-7312  History of unacceptable prior treatment relate...   
15        TCGA-PN-A8M9                                  Genotype mismatch   
16        TCGA-AG-3612  History of unacceptable prior treatment relate...   
17        TCGA-XK-AAJ3  History of unacceptable prior treatment relate...   
18        TCGA-WB-A820  History of unacceptable prior treatment relate...   
19        TCGA-L5-A88T  History of unacceptable prior treatment relate...   
20        TCGA-EY-A1G8  History of unacceptable prior treatment relate...   
21        TCGA-D1-A3JP  History of unacceptable prior treatment relate...   
22        TCGA-BG-A221  History of unacceptable prior treatment relate...   
23        TCGA-BG-A220  History of unacceptable prior treatment relate...   
24        TCGA-B5-A0KB  History of unacceptable prior treatment relate...   
25        TCGA-AX-A3G3  History of unacceptable prior treatment relate...   
26        TCGA-AX-A2HF  History of unacceptable prior treatment relate...   
27        TCGA-AX-A2H8  History of unacceptable prior treatment relate...   
28        TCGA-AX-A1CR  History of unacceptable prior treatment relate...   
29        TCGA-AX-A1CP  History of unacceptable prior treatment relate...   
..                 ...                                                ...   
182       TCGA-BR-4196                                     Duplicate case   
183       TCGA-14-1824                                  Genotype mismatch   
184       TCGA-02-2488                                  Genotype mismatch   
185       TCGA-06-0748                                  Genotype mismatch   
186       TCGA-14-1036                                  Genotype mismatch   
187       TCGA-14-0784                                  Genotype mismatch   
188       TCGA-33-4579                      Tumor tissue origin incorrect   
189       TCGA-06-0131                           Subject withdrew consent   
190       TCGA-BR-4282                                     Duplicate case   
191       TCGA-BR-4205                                     Duplicate case   
192       TCGA-E2-A1IP                                  Genotype mismatch   
193       TCGA-A6-2670                           Subject withdrew consent   
194       TCGA-AF-3912                           Subject withdrew consent   
195       TCGA-32-2498                           Subject withdrew consent   
196       TCGA-08-0384                                  Genotype mismatch   
197       TCGA-AN-A0FE                           Subject identity unknown   
198       TCGA-AN-A0FG                           Subject identity unknown   
199       TCGA-24-0981                                     Duplicate case   
200       TCGA-02-0002                                  Genotype mismatch   
201       TCGA-02-0117                                  Genotype mismatch   


and here we explicitly call just the second query, resulting in 329 patients:

In [12]:
bq.Query(createCohort_and_checkAnnotations.select_on_clinical).results().to_dataframe()

ParticipantBarcode vital_status  days_to_last_known_alive  \
0         TCGA-BH-A0DK        Alive                       423   
1         TCGA-BH-A0DZ        Alive                       495   
2         TCGA-A7-A5ZW        Alive                       326   
3         TCGA-A7-A5ZX        Alive                       336   
4         TCGA-A2-A1G0        Alive                       616   
5         TCGA-B6-A3ZX         Dead                      1152   
6         TCGA-A2-A1G6        Alive                       501   
7         TCGA-AR-A2LJ        Alive                      2632   
8         TCGA-B6-A0X1         Dead                      7455   
9         TCGA-AR-A2LO        Alive                      1198   
10        TCGA-AR-A2LM        Alive                      1935   
11        TCGA-A2-A04X        Alive                      1686   
12        TCGA-E9-A6HE        Alive                       847   
13        TCGA-PE-A5DE        Alive                      2645   
14        TCGA-A2-A04Q        Alive                      2385   
15        TCGA-EW-A1IX        Alive                      1208   
16        TCGA-EW-A1IY        Alive                       258   
17        TCGA-A2-A04P         Dead                       548   
18        TCGA-A2-A04U        Alive                      2654   
19        TCGA-A2-A04W        Alive                      3102   
20        TCGA-E2-A10A        Alive                      1229   
21        TCGA-E2-A10F        Alive                       878   
22        TCGA-BH-A0DG        Alive                      2041   
23        TCGA-A2-A0EX        Alive                       752   
24        TCGA-AR-A24R        Alive                      3430   
25        TCGA-AR-A24P        Alive                        84   
26        TCGA-AR-A24U        Alive                      3128   
27        TCGA-AR-A24T        Alive                      3202   
28        TCGA-AR-A24K        Alive                      1548   
29        TCGA-AR-A24O        Alive                      3607   
..                 ...          ...                       ...   
299       TCGA-AO-A0J4        Alive                      1587   
300       TCGA-AO-A0J5         Dead                       792   
301       TCGA-AO-A0J2        Alive                       997   
302       TCGA-E9-A3QA        Alive                       918   
303       TCGA-AO-A0JM        Alive                      2184   
304       TCGA-AO-A0JB        Alive                      1542   
305       TCGA-B6-A409         Dead                       573   
306       TCGA-B6-A400        Alive                       215   
307       TCGA-B6-A402        Alive                      2134   
308       TCGA-B6-A0I6         Dead                       991   
309       TCGA-B6-A0I8         Dead                       749   
310       TCGA-E2-A15D        Alive                       526   
311       TCGA-E2-A15A        Alive                       710   
312       TCGA-E2-A15S        Alive                       428   
313       TCGA-E2-A159        Alive                       762   
314       TCGA-E2-A158        Alive                       450   
315       TCGA-E2-A150        Alive                      1935   
316       TCGA-AR-A2LR        Alive                      1742   
317       TCGA-A2-A0CL        Alive                      3015   
318       TCGA-A2-A0CZ        Alive                      1616   
319       TCGA-C8-A12Z        Alive                       382   
320       TCGA-C8-A12Y        Alive                      1476   
321       TCGA-C8-A12O        Alive                       385   
322       TCGA-AO-A1KO        Alive                       622   
323       TCGA-AQ-A1H3        Alive                       989   
324       TCGA-BH-A1ES         Dead                      3462   
325       TCGA-BH-A1EV         Dead                       365   
326       TCGA-BH-A1EW         Dead                      1694   
327       TCGA-AO-A03M        Alive                      1866   
328       TCGA-AO-A03L        Alive    

and finally we call the main query:

In [13]:
bq.Query(createCohort_and_checkAnnotations).results().to_dataframe()

ParticipantBarcode
0         TCGA-3C-AALI
1         TCGA-4H-AAAK
2         TCGA-5L-AAT0
3         TCGA-A1-A0SH
4         TCGA-A1-A0SJ
5         TCGA-A1-A0SN
6         TCGA-A1-A0SP
7         TCGA-A1-A0SQ
8         TCGA-A2-A04P
9         TCGA-A2-A04Q
10        TCGA-A2-A04R
11        TCGA-A2-A04U
12        TCGA-A2-A04V
13        TCGA-A2-A04W
14        TCGA-A2-A04X
15        TCGA-A2-A0CL
16        TCGA-A2-A0CM
17        TCGA-A2-A0CV
18        TCGA-A2-A0CZ
19        TCGA-A2-A0D2
20        TCGA-A2-A0D3
21        TCGA-A2-A0D4
22        TCGA-A2-A0EX
23        TCGA-A2-A0SX
24        TCGA-A2-A0T3
25        TCGA-A2-A0T5
26        TCGA-A2-A0T6
27        TCGA-A2-A0YE
28        TCGA-A2-A0YJ
29        TCGA-A2-A0YL
..                 ...
297       TCGA-GM-A2DL
298       TCGA-GM-A3XG
299       TCGA-GM-A3XL
300       TCGA-GM-A3XN
301       TCGA-HN-A2OB
302       TCGA-JL-A3YW
303       TCGA-JL-A3YX
304       TCGA-LD-A66U
305       TCGA-LL-A5YN
306       TCGA-LL-A5YO
307       TCGA-LL-A5YP
308       TCGA-LL-A6FR
309       TCGA-LL-A7SZ
310       TCGA-MS-A51U
311       TCGA-OL-A5D8
312       TCGA-OL-A5RV
313       TCGA-OL-A5RW
314       TCGA-OL-A66I
315       TCGA-OL-A66O
316       TCGA-OL-A6VO
317       TCGA-OL-A6VQ
318       TCGA-OL-A6VR
319       TCGA-PE-A5DE
320       TCGA-PL-A8LX
321       TCGA-PL-A8LY
322       TCGA-PL-A8LZ
323       TCGA-S3-A6ZH
324       TCGA-S3-AA14
325       TCGA-XX-A899
326       TCGA-Z7-A8R6

[327 rows x 1 columns]

Note that we didn't need to call each sub-query individually, we could have just called the main query and gotten the same result.  As you can see, two patients that met the clinical select criteria (which returned 329 patients) were excluded from the final result (which returned 327 patients).

Before we leave off, here are a few useful tricks for working with BigQuery in Cloud Datalab:
+ if you want to see the raw SQL, you can just build the query and then print it out (this might be useful, for example, in debugging a query -- you can copy and paste the SQL directly into the BigQuery Web UI);
+ if you want to see how much data and which tables are going to be touched by this data, you can use the "dry run" option.  (Notice the "cacheHit" flag -- if you have recently done a particular query, you will not be charged to repeat it since it will have been cached.)

In [14]:
q = bq.Query(createCohort_and_checkAnnotations)
q

SELECT
  c.ParticipantBarcode AS ParticipantBarcode
FROM (
  SELECT
    a.categoryName,
    a.classificationName,
    a.ParticipantBarcode,
    c.ParticipantBarcode,
  FROM ( (SELECT
  ParticipantBarcode,
  annotationCategoryName AS categoryName,
  annotationClassification AS classificationName
FROM
  [isb-cgc:tcga_201510_alpha.Annotations]
WHERE
  ( itemTypeName="Patient"
    AND (annotationCategoryName="History of unacceptable prior treatment related to a prior/other malignancy"
      OR annotationClassification="Redaction" ) )
GROUP BY
  ParticipantBarcode,
  categoryName,
  classificationName) ) AS a
  OUTER JOIN EACH 
       ( (SELECT
  ParticipantBarcode,
  vital_status,
  days_to_last_known_alive,
  ethnicity,
  histological_type,
  menopause_status,
  race
FROM
  [isb-cgc:tcga_201510_alpha.Clinical_data]
WHERE
  ( Study="BRCA"
    AND age_at_initial_pathologic_diagnosis<=50
    AND gender="FEMALE" )) ) AS c
  ON
    a.ParticipantBarcode = c.ParticipantBarcode
  WHERE
    (a.ParticipantBarcode IS NOT NULL
      OR c.ParticipantBarcode IS NOT NULL)
  ORDER BY
    a.classificationName,
    a.categoryName,
    a.ParticipantBarcode,
    c.ParticipantBarcode )
WHERE
  ( a.categoryName IS NULL
    AND a.classificationName IS NULL
    AND c.ParticipantBarcode IS NOT NULL )
ORDER BY
  c.ParticipantBarcode

In [15]:
q.execute_dry_run()

{u'cacheHit': True,
 u'referencedTables': [{u'datasetId': u'tcga_201510_alpha',
   u'projectId': u'isb-cgc',
   u'tableId': u'Clinical_data'},
  {u'datasetId': u'tcga_201510_alpha',
   u'projectId': u'isb-cgc',
   u'tableId': u'Annotations'}],
 u'totalBytesBilled': u'0',
 u'totalBytesProcessed': u'1467388'}